### PR TITLE
Calculate shouldExpandNode on JSONNestedNode.render

### PR DESCRIPTION
**Pull request by gnestor**
Closes https://github.com/alexkuz/react-json-tree/issues/60

### CODE REVIEW

**Comment by gnestor** [comment]
Here's a test component:

```jsx
import React from 'react';
import JSONTree from 'react-json-tree';

export class JSONFilterableTree extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      filter: ''
    };
  }

  render() {
    let data = this.props.data;
    if (this.state.filter) {
      data = Object.keys(this.props.data).reduce((result, key) => {
        if (JSON.stringify(this.props.data[key]).includes(this.state.filter)) result[key] = this.props.data[key];
        return result;
      }, {});
    }
    return (
      <div
        style={{
          position: 'relative'
        }}
      >
        <input
          value={this.state.filter}
          onChange={(event) => {
            this.setState({filter: event.target.value});
          }}
          style={{
            position: 'absolute',
            top: 0,
            right: 0,
            width: '33%',
            maxWidth: 150,
            zIndex: 10,
            fontSize: 13,
            paddingHorizontal: 4,
            paddingVertical: 2
          }}
          type="text"
          placeholder="Filter..."
        />
        <JSONTree data={data} />
      </div>
    );
  }

}
```

File: src/JSONNestedNode.js [hunk 1]
```diff
@@ -114,7 +114,12 @@ export default class JSONNestedNode extends React.Component {
...
       expandable
     } = this.props;
-    const expanded = this.state.expanded;
+    let expanded = this.state.expanded;
+    if (this.state.expanded === false) {
+      expanded = this.props.shouldExpandNode && !this.props.isCircular ?
+      this.props.shouldExpandNode(this.props.keyPath, this.props.data, this.props.level) :
+      false;
+    }
     const renderedChildren = expanded || (hideRoot && this.props.level === 0) ?
```
Comment (alexkuz) [inline src/JSONNestedNode.js]: I think it's better to move that to `componentWillReceiveProps`, like this:
```js
const getStateFromProps = props => ({
  ...  // calculate `expanded`
});

...

constructor(props) {
  super(props);
  this.state = getStateFromProps(props);
}

componentWillReceiveProps(nextProps) {
  if (['shouldExpandNode', 'isCircular', 'keyPath', 'data', 'level'].find(k => nextProps[k] !== this.props[k]) {
    this.setState(getStateFromProps(nextProps));
  }
}
```

Comment (gnestor) [inline src/JSONNestedNode.js]: Ya, better to keep it DRY. I tested and this is working 👍

**Comment by gnestor** [comment]
@alexkuz I made the changes you suggested. I think this is ready to merge 👍

**Comment by cephirothdy2j** [comment]
+1 on this; I'm waiting on the same fix. Will this be merged and released any time soon?

**Comment by alexkuz** [comment]
Yes, I'll take a look soon, sorry for delay.

File: src/JSONNestedNode.js [hunk 2]
```diff
@@ -58,6 +58,17 @@ function renderChildNodes(props, from, to) {
...
 }
 
+function getStateFromProps(props) {
+  // calculate individual node expansion if necessary
+  const expanded = props.shouldExpandNode && !props.isCircular ?
+    props.shouldExpandNode(props.keyPath, props.data, props.level) :
+    false;
+  return {
+    expanded,
+    createdChildNodes: false
```
Comment (alexkuz) [inline src/JSONNestedNode.js]: What is this for? Seems like it's not used anywhere.

**Comment by alexkuz** [comment]
@gnestor Thank you! Everything is fine, except for that seemingly redundant state value.

**Comment by alexkuz** [comment]
Oh, wait, it was there before. Nevermind then :)

**Comment by zalmoxisus** [comment]
Could we move the changes from this pr under a flag (prop)?

As mentioned in #66, it breaks tree inspection for both Redux DevTools Inspector and Log Monitor:

wbhsmo0vt8

**Comment by gnestor** [comment]
What if we changed the following:

```js
componentWillReceiveProps(nextProps) {
  if (['shouldExpandNode', 'isCircular', 'keyPath', 'data', 'level'].find(k =>
    nextProps[k] !== this.props[k])
  ) {
    this.setState(getStateFromProps(nextProps));
  }
}
```

to:

```js
componentWillReceiveProps(nextProps) {
  if (getStateFromProps(nextProps) !== getStateFromProps(this.props)) {
    this.setState(getStateFromProps(nextProps));
  }
}
```

In its current form, `componentWillReceiveProps` is over-eager to recalculate `expanded` state, but this change should correct that (because it appears that the only value that is changing in the above screencap is `data`). @zalmoxisus Can you test whether this fixes the issue in redux-devtools-inspector? And whether it has any effect on performance?

**Comment by zalmoxisus** [comment]
@gnestor it doesn't help, but `if (false)` helps 😂

You can clone [`redux-devtools-inspector`](https://github.com/alexkuz/redux-devtools-inspector) and run `yarn && yarn start` to open the demo.

Not sure if solving that specific issue when an action is selected (as in my screenshot) would be enough. According to #66, users got used that, even when selecting a different action, the state tree remains expanded. If so, we could just add a prop to enable the introduced updates. @alexkuz what do you think?

**Comment by alexkuz** [comment]
@zalmoxisus try `v0.10.3` (or RDI `v0.11.2`), it should work now.

As @gnestor proposed, now it just checks if `state.expanded` was changed.

**Comment by zalmoxisus** [comment]
@alexkuz yes, it works. Thanks a lot for the fix!

**Pull request merged**